documentation/ControlFlowGraph.md

```diff
@@ -0,0 +1,157 @@
+single entry point and a single exit point, and a collection of
+directed flow _edges_ (or _arcs_) denoting all possible transition of
+control flow that may take place during execution from the end of
+one basic block to the beginning of another (or itself).
+
+A block that has multiple distinct successors in the flow of control
+must end with an action that selects its successor.
+
+The sequence of actions that constitutes a basic block may
+include references to user and library procedures.
+Subprogram calls with implicit control flow afterwards, namely
+alternate returns and `END=`/`ERR=` labels on input/output,
+will be lowered in translation to a representation that materializes
+that control flow into something similar to a computed `GO TO` or
+C language `switch` statement.
+
+For convenience in optimization and to simplify the implementation of
+data flow confluence functions, we may choose to maintain the
+property that each flow arc is the sole outbound arc emanating from
+its originating block, the sole inbound arc arriving at its destination,
+or both.
+Empty blocks would inserted to "split" arcs when necessary to maintain this
+invariant property.
+
+Fortran subprograms (other than internal subprograms) can have multiple
+entry points by using the obsolescent `ENTRY` statement.
+We will implement such subprograms by constructing a union
+of their dummy argument lists and using it as part of the definition
+of a new subroutine or function that can be called by each of
+the entry points, which are then all converted into wrapper routines that
+pass a selector value as an additional argument to drive a `switch` on entry
+to the new subprogram.
+
+This transformation ensures that every subprogram's control
+flow graph has a well-defined `START` node.
+
+Statement labels can be used in Fortran on any statement, but only
+the labels that decorate legal destinations of `GO TO` statements
+need to be implemented in the control flow graph.
+Specifically, non-executable statements like `DATA`, `NAMELIST`, and
+`FORMAT` statements will be extracted into data initialization
+records before or during the construction of the control flow
+graph, and will survive only as synonyms for `CONTINUE`.
+
+Nests of multiple labeled `DO` loops that terminate on the same
+label will be have that label rewritten so that `GO TO` within
+the loop nest will arrive at the copy that most closely nests
+the context.
+The Fortran standard does not require us to do this, but XLF
+(at least) works this way.
+
+### Expressions and Statements (Operations and Actions)
+Expressions are trees, not DAGs, of intrinsic operations,
+resolved function references, constant literals, and
+data designators.
+
+Expression nodes are represented in the compiler in a type-safe manner.
+There is a general `Expression` class template that is parameterized
+over type classification (e.g., `INTEGER`, `REAL`, &c.) and, for intrinsic
+types, kind type parameter values.
```

Comment:
I think this is out of date. It's now IntegerExpr, RealExpr, ComplexExpr, CharacterExpr.

Reply:
Right, will update.